### ISSUE TITLE
@JsonbPropertyOrder implementation is according to specs but is not intuitive and too limited

### ISSUE THREAD

**Issue by steappe** [eclipse-ee4j/yasson#159]
Let's have a POJO annotated with JSONB:

```
@JsonbPropertyOrder(
    {
        "from",
        "count"
    }
)
public class Range {

    @JsonbProperty("from")
    public final int fromIndex;

    @JsonbProperty("count")
    public final int numberOfItems;

    @JsonbCreator
    public Range(
        @JsonbProperty("from") int fromIndex,
        @JsonbProperty("count") int numberOfItems
    ) {
        this.fromIndex = fromIndex;
        this. numberOfItems = numberOfItems;
    }
}
```
It is serialized as:
```
{
  "count": count,
  "from": from
}
```
This is correct, because according to the specifications:

> Names must correspond to original names defined in Java class before any customization applied.

But in a first place I would have expected the following serialization:
```
{
  "from": from,
  "count": count
}
```

If I change the property order annotation to:
```
@JsonbPropertyOrder(
    {
        "fromIndex",
        "numberOfItems"
    }
)
```

It is correctly serialized as:
```
{
  "from": from,
  "count": count
}
```

This behavior is not intuitive and makes it impossible to define constant properties that can be used both in `@JsonbProperty` and `@JsonbPropertyOrder`, as follows:
```
@JsonbPropertyOrder(
    {
        Range.PROP_FROM,
        Range.PROP_COUNT
    }
)
public class Range {

    static public final String PROP_FROM = "from";
    static public final String PROP_COUNT = "count";

    @JsonbProperty(PROP_FROM)
    public final int fromIndex;

    @JsonbProperty(PROP_COUNT)
    public final int numberOfItems;

    @JsonbCreator
    public Range(
        @JsonbProperty(PROP_FROM) int fromIndex,
        @JsonbProperty(PROP_COUNT) int numberOfItems
    ) {
        this.fromIndex = fromIndex;
        this. numberOfItems = numberOfItems;
    }
}
```
The code above will just not define the property order I would expect.
I can't see any workaround.

Edit:
According to the JSON binding specifications:

> 4.2 Customizing Property Order
> To customize the order of serialized properties, JSON Binding provides javax.json.bind.config.PropertyOrderStrategy class.
> Class javax.json.bind.config.PropertyOrderStrategy provides the most common property order strategies.
>  LEXICOGRAPHICAL  ANY
>  REVERSE
> The detailed description of property order strategies can be found in javadoc. The way to set custom property order strategy is to use
> javax.json.bind.JsonbConfig::withPropertyOrderStrategy method.
> To customize the order of serialized properties only for one specific type, JSON Binding provides javax.json.bind.annotation.JsonbPropertyOrder annotation. Order specified by JsonbPropertyOrder annotation overrides order specified by PropertyOrderStrategy.
> **The order is applied to already renamed properties as stated in 4.1.**

And, according to the Java DOC:

> The value parameter of the `@JsonbPropertyOrder`
> Order in which properties are serialized. Names must correspond to original names defined in Java class before any customization applied.

In short, the property order strategy is applied after the customization, and the value specified by the property order annotation is applied before the customization. On the one hand this is not consistent; on the other hand, we can't use constant properties for both `@JsonbProperty` and `@JsonbPropertyOrder`. IMHO, something needs to be added in the JSON binding specs to fix this.
